sql/core/src/main/scala/org/apache/spark/sql/execution/python/ArrowEvalPythonExec.scala

```diff
@@ -89,9 +89,35 @@ case class ArrowEvalPythonExec(udfs: Seq[PythonUDF], resultAttrs: Seq[Attribute]
 
     columnarBatchIter.flatMap { batch =>
       val actualDataTypes = (0 until batch.numCols()).map(i => batch.column(i).dataType())
-      assert(outputTypes == actualDataTypes, "Invalid schema from pandas_udf: " +
-        s"expected ${outputTypes.mkString(", ")}, got ${actualDataTypes.mkString(", ")}")
+      assert(plainSchema(outputTypes) == actualDataTypes,
+        "Incompatible schema from pandas_udf: " +
+          s"expected ${outputTypes.mkString(", ")}, got ${actualDataTypes.mkString(", ")}")
       batch.rowIterator.asScala
     }
   }
+
+  private def plainSchema(schema: Seq[DataType]): Seq[DataType] =
+    schema.map(v => plainSchema(v, false)).toList
+
+  /** Erase User-Defined Types and returns the plain Spark StructType instead.
+   *
+   * @note
+   * PyArrow returns `ArrayType` with `containsNull=true`
```

Comment:
What does this means? What if the return type is `@pandas_udf(ArrayType(xxx, False))`?

Reply:
Is PyArrow always returning `ArrayType` with `containsNull=true`?

Why? Is is a feature of PyArrow or it is hard-coded somewhere in pyspark?

Reply:
Right, this is auto inferred from spark's pandas `from_arrow_type()` function.

https://github.com/apache/spark/blob/9b875ceada60732899053fbd90728b4944d1c03d/python/pyspark/sql/pandas/types.py#L121-L125